### PR TITLE
[WIP] adding Kandinsky training scripts

### PR DESCRIPTION
**Pull request by yiyixuxu**
The author of this PR is @cene555 and the Kandinsky team. Based on the [original PR](https://github.com/huggingface/diffusers/pull/4051),  I'm working on integrating the Kandinsky training scripts into diffusers in this PR! 

## 🚨🚨🚨 **Note**: The **main** author of this PR is @cene555 and the Kandinsky team. In this PR we mainly just make some small modifications to the training script provided by the authors from [original PR](https://github.com/huggingface/diffusers/pull/4051) so that it's consistent with all our other training scripts. Thanks a million for the contribution @cene555  🚨🚨🚨

Authors of this PR:
[Arseniy Shakhmatov](https://github.com/cene555) 
[Anton Razzhigaev](https://github.com/razzant) 
[Aleksandr Nikolich](https://github.com/AlexWortega) 
[Igor Pavlov](https://github.com/boomb0om) 
[Andrey Kuznetsov](https://github.com/kuznetsoffandrey)
[Denis Dimitrov](https://github.com/denndimitrov)

Note: 
will include dreambooth in a separate PR

### CODE REVIEW

File: src/diffusers/models/prior_transformer.py [hunk 1]
```diff
@@ -32,7 +33,7 @@ class PriorTransformerOutput(BaseOutput):
     predicted_image_embedding: torch.FloatTensor
 
 
-class PriorTransformer(ModelMixin, ConfigMixin):
+class PriorTransformer(ModelMixin, ConfigMixin, UNet2DConditionLoadersMixin):
```
Comment (yiyixuxu) [inline src/diffusers/models/prior_transformer.py:36]: I'm a little bit surprised to see `UNet2DConditionLoadersMixin` works out of box for `PriorTransformer`:)

Comment (sayakpaul) [reply src/diffusers/models/prior_transformer.py:36]: Just `diffusers` things 😎

File: examples/kandinsky2_2/text_to_image/README.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
+# Kandinsky2.2 text-to-image fine-tuning
+
+Kandinsky 2.2 includes a prior pipeline that generates image embeddings from text prompts, and a decoder pipeline to that generate the output image based on the image embeddings. We provide `train_text_to_image_prior.py` and `train_text_to_image_decoder.py` scripts to show you how to fine-tune the kandinsky prior and decoder models seperately based on your own dataset. To achieve the best resutls, you should fine-tune both your prior and decoder models.
```
Comment (sayakpaul) [inline examples/kandinsky2_2/text_to_image/README.md:3]: Let's leave a link to the Kandinsky paper or the official repo or something that was officially released by the authors?

Comment (sayakpaul) [reply examples/kandinsky2_2/text_to_image/README.md:3]:
```suggestion
Kandinsky 2.2 includes a prior pipeline that generates image embeddings from text prompts, and a decoder pipeline that generates the output image based on the image embeddings. We provide `train_text_to_image_prior.py` and `train_text_to_image_decoder.py` scripts to show you how to fine-tune the Kandinsky prior and decoder models separately based on your own dataset. To achieve the best results, you should fine-tune **_both_** your prior and decoder models.
```

Let's maintain "Kandinsky" throughout.

File: examples/kandinsky2_2/text_to_image/README.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+pipe_prior = DiffusionPipeline.from_pretrained(output_dir, torch_dtype=torch.float16)
+prior_components = {"prior_" + k: v for k,v in pipe_prior.components.items()}
+pipe = AutoPipelineForText2Image.from_pretrained("kandinsky-community/kandinsky-2-2-decoder", **prior_components, torch_dtype=torch.float16)
+
+pipe.enable_model_cpu_offload()
+prompt='A robot pokemon, 4k photo'
+images = pipe(prompt=prompt, negative_prompt=negative_prompt).images
+images[0]
+```
```
Comment (sayakpaul) [inline examples/kandinsky2_2/text_to_image/README.md:177]: Can we also make a note on how to perform inference when someone fine-tunes both the prior and the decoder? Of course, it's obvious to us as to what needs to be done but it might not be for all.

File: examples/kandinsky2_2/text_to_image/README.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+### Training with xFormers:
+
+You can enable memory efficient attention by [installing xFormers](https://huggingface.co/docs/diffusers/main/en/optimization/xformers) and passing the `--enable_xformers_memory_efficient_attention` argument to the script.
+
+xFormers training is not available for Prior model fine-tune.
+
+**Note**:
+
+According to [this issue](https://github.com/huggingface/diffusers/issues/2234#issuecomment-1416931212), xFormers `v0.0.16` cannot be used for training in some GPUs. If you observe that problem, please install a development version as indicated in that comment.
```
Comment (sayakpaul) [inline examples/kandinsky2_2/text_to_image/README.md:317]: We have LoRA too. Let's include a section on LoRA as well in the README.

File: examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py [hunk 5]
```diff
@@ -0,0 +1,820 @@
...
+        if name.startswith("mid_block"):
+            hidden_size = unet.config.block_out_channels[-1]
+        elif name.startswith("up_blocks"):
+            block_id = int(name[len("up_blocks.")])
+            hidden_size = list(reversed(unet.config.block_out_channels))[block_id]
+        elif name.startswith("down_blocks"):
+            block_id = int(name[len("down_blocks.")])
+            hidden_size = unet.config.block_out_channels[block_id]
+
+        lora_attn_procs[name] = LoRAAttnAddedKVProcessor(
```
Comment (sayakpaul) [inline examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py:414]: Hmm. Seems like we don't have a faster implementation of this processor. Based on the usage of the training scripts, I think we can monitor and added them later if needed. WDYT?

File: examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py [hunk 6]
```diff
@@ -0,0 +1,820 @@
...
+
+                target = noise
+
+                # Predict the noise residual and compute loss
+                added_cond_kwargs = {"image_embeds": image_embeds}
+
+                model_pred = unet(noisy_latents, timesteps, None, added_cond_kwargs=added_cond_kwargs).sample[:, :4]
+
+                if args.snr_gamma is None:
+                    loss = F.mse_loss(model_pred.float(), target.float(), reduction="mean")
```
Comment (sayakpaul) [inline examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py:651]: IIRC Kandinsky models weren't trained using the epsilon prediction objective. If so, does this still lead to reasonable results?

**Review by sayakpaul — APPROVED** [review]
Amazing work! 

Questions / comments:

* Do we not LoRA fine-tune the text encoders as typically done in the SD world? 
* Let's add tests! Pinging @DN6 for a second opinion here. 
* Let's port over the README to our training documentation too so that it reflects on https://huggingface.co/docs/diffusers/main/en/index.

File: examples/kandinsky2_2/text_to_image/README.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+  --validation_prompts="A robot pokemon, 4k photo" \
+  --report_to="wandb" \
+  --push_to_hub \
+  --output_dir="kandi2-decoder-pokemon-model" 
+```
+<!-- accelerate_snippet_end -->
+
+
+To run on your own training files prepare the dataset according to the format required by `datasets`, you can find the instructions for how to do that in this [document](https://huggingface.co/docs/datasets/v2.4.0/en/image_load#imagefolder-with-metadata).
+If you wish to use custom loading logic, you should modify the script, we have left pointers for that in the training script.
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:83]:
```suggestion
To train on your own training files, prepare the dataset according to the format required by `datasets`. You can find the instructions for how to do that in the [ImageFolder with metadata](https://huggingface.co/docs/datasets/en/image_load#imagefolder-with-metadata) guide.
If you wish to use custom loading logic, you should modify the script and we have left pointers for that in the training script.
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+pipe = AutoPipelineForText2Image.from_pretrained("kandinsky-community/kandinsky-2-2-decoder", unet=unet, torch_dtype=torch.float16)
+pipe.enable_model_cpu_offload()
+
+image = pipe(prompt="A robot pokemon, 4k photo").images[0]
+image.save("robot-pokemon.png")
+```
+
+#### Fine-tune prior 
+
+You can fine-tune kandinsky prior model with `train_text_to_image_prior.py` script. Note that we currently do not support `--gradient_checkpointing` for prior model fine-tune.
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:137]:
```suggestion
You can fine-tune the Kandinsky prior model with `train_text_to_image_prior.py` script. Note that we currently do not support `--gradient_checkpointing` for prior model fine-tuning.
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+  --lr_scheduler="constant" --lr_warmup_steps=0 \
+  --validation_prompts="A robot pokemon, 4k photo" \
+  --report_to="wandb" \
+  --push_to_hub \
+  --output_dir="kandi2-prior-pokemon-model" 
+```
+<!-- accelerate_snippet_end -->
+
+
+To perform inference with the fune-tuned prior model, you will need to first create a prior pipeline by passing the `output_dir` to `DiffusionPpeline`, and then you can create a `KandinskyV22CombinedPipeline` with a pretrained or fine-tuned decoder checkpoint, along with all the modules of the prior pipeline you just created.
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:163]:
```suggestion
To perform inference with the fine-tuned prior model, you will need to first create a prior pipeline by passing the `output_dir` to `DiffusionPipeline`. Then create a `KandinskyV22CombinedPipeline` from a pretrained or fine-tuned decoder checkpoint along with all the modules of the prior pipeline you just created. 
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+  --push_to_hub \
+  --output_dir="kandi2-decoder-pokemon-model" 
+```
+
+
+#### Training with Min-SNR weighting
+
+We support training with the Min-SNR weighting strategy proposed in [Efficient Diffusion Training via Min-SNR Weighting Strategy](https://arxiv.org/abs/2303.09556) which helps to achieve faster convergence
+by rebalancing the loss. In order to use it, one needs to set the `--snr_gamma` argument. The recommended
+value when using it is 5.0.
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:210]:
```suggestion
We support training with the Min-SNR weighting strategy proposed in [Efficient Diffusion Training via Min-SNR Weighting Strategy](https://arxiv.org/abs/2303.09556) which helps achieve faster convergence
by rebalancing the loss. Enable the `--snr_gamma` argument and set it to the recommended
value of 5.0.
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+In a nutshell, LoRA allows adapting pretrained models by adding pairs of rank-decomposition matrices to existing weights and **only** training those newly added weights. This has a couple of advantages:
+
+- Previous pretrained weights are kept frozen so that model is not prone to [catastrophic forgetting](https://www.pnas.org/doi/10.1073/pnas.1611835114).
+- Rank-decomposition matrices have significantly fewer parameters than original model, which means that trained LoRA weights are easily portable.
+- LoRA attention layers allow to control to which extent the model is adapted toward new training images via a `scale` parameter.
+
+[cloneofsimo](https://github.com/cloneofsimo) was the first to try out LoRA training for Stable Diffusion in the popular [lora](https://github.com/cloneofsimo/lora) GitHub repository.
+
+With LoRA, it's possible to fine-tune Kandinsky2.2 on a custom image-caption pair dataset
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:225]:
```suggestion
With LoRA, it's possible to fine-tune Kandinsky 2.2 on a custom image-caption pair dataset
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+on consumer GPUs like Tesla T4, Tesla V100.
+
+### Training
+
+First, you need to set up your development environment as is explained in the [installation section](#installing-the-dependencies). Make sure to set the `MODEL_NAME` and `DATASET_NAME` environment variables. Here, we will use [Kandinsky2.2](https://huggingface.co/kandinsky-community/kandinsky-2-2-decoder) and the [Pokemons dataset](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions).
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:230]:
```suggestion
First, you need to set up your development environment as explained in the [installation](#installing-the-dependencies). Make sure to set the `MODEL_NAME` and `DATASET_NAME` environment variables. Here, we will use [Kandinsky 2.2](https://huggingface.co/kandinsky-community/kandinsky-2-2-decoder) and the [Pokemons dataset](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions).  
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+```
+
+**___Note: When using LoRA we can use a much higher learning rate compared to non-LoRA fine-tuning. Here we use *1e-4* instead of the usual *1e-5*. Also, by using LoRA, it's possible to run above scripts in consumer GPUs like T4 or V100.___**
+
+
+### Inference
+
+#### Inference using fine-tuned LoRA checkpoint for decoder
+
+Once you have trained a kandinsky decoder model using above command, the inference can be done simply using the `AutoPipelineForText2Image` after loading the trained LoRA weights.  You need to pass the `output_dir` for loading the LoRA weights which, in this case, is `kandi22-decoder-pokemon-lora`.
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:277]:
```suggestion
Once you have trained a Kandinsky decoder model using the above command, inference can be done with the `AutoPipelineForText2Image` after loading the trained LoRA weights.  You need to pass the `output_dir` for loading the LoRA weights, which in this case is `kandi22-decoder-pokemon-lora`.
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+pipe = AutoPipelineForText2Image.from_pretrained("kandinsky-community/kandinsky-2-2-decoder", torch_dtype=torch.float16)
+pipe.unet.load_attn_procs(output_dir)
+pipe.enable_model_cpu_offload()
+
+prompt='A robot pokemon, 4k photo'
+image = pipe(prompt=prompt).images[0]
+image.save("robot_pokemon.png")
+```
+
+#### Inference using fine-tuned LoRA checkpoint for Prior
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:293]:
```suggestion
#### Inference using fine-tuned LoRA checkpoint for prior
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+
+```python
+from diffusers import AutoPipelineForText2Image
+import torch
+
+pipe = AutoPipelineForText2Image.from_pretrained("kandinsky-community/kandinsky-2-2-decoder", torch_dtype=torch.float16)
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:299]:
```suggestion
pipe = AutoPipelineForText2Image.from_pretrained("kandinsky-community/kandinsky-2-2-prior", torch_dtype=torch.float16)
```

File: examples/kandinsky2_2/text_to_image/README.md [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+image = pipe(prompt=prompt).images[0]
+image.save("robot_pokemon.png")
+image
+```
+
+### Training with xFormers:
+
+You can enable memory efficient attention by [installing xFormers](https://huggingface.co/docs/diffusers/main/en/optimization/xformers) and passing the `--enable_xformers_memory_efficient_attention` argument to the script.
+
+xFormers training is not available for Prior model fine-tune.
```
Comment (stevhliu) [inline examples/kandinsky2_2/text_to_image/README.md:313]:
```suggestion
xFormers training is not available for fine-tuning the prior model.
```

**Review by stevhliu — APPROVED** [review]
Very nice, thanks for writing this training guide up! 😄

File: examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py (hunk 6 above)
Comment (yiyixuxu) [reply examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py:651]: I think they are just following with dalle-2 practice, which is trained with the same objective. it leads to better results based on the paper 

"Instead of using the -prediction formulation from Ho et al. [25], we find it better to train our model to predict the unnoised zi directly, and use a mean-squared error loss on this prediction"

File: examples/kandinsky2_2/text_to_image/README.md (hunk 15 above)
Comment (yiyixuxu) [reply examples/kandinsky2_2/text_to_image/README.md:298]: Actually, here we want to create a combined pipeline, so will have to use the decoder checkpoint for that
once we have the `pipe` as the combined pipeline, we then access the prior with `pipe.prior_prior`

File: examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py (hunk 6 above)
Comment (yiyixuxu) [reply examples/kandinsky2_2/text_to_image/train_text_to_image_lora_decoder.py:651]: Actually I'm not sure what's the objective used in decoder training, I looked back into the dalle-2 paper, and I think it only mentioned that it predicts samples directly for prior

**Comment by yiyixuxu** [comment]
Is it okay to merge now? Maybe @sayakpaul can take a look again 
I don't think we need to add this to doc; I added a link to this folder on the text-to-image training doc page instead

**Comment by yiyixuxu** [comment]
@sayakpaul about your questions
 
> Do we not LoRA fine-tune the text encoders as typically done in the SD world?

kandinsky does not have a pure text-conditioned diffusion process - it decodes image embedding instead. it does use a text encoder but I'm not sure how important it is to fine-tune it. Maybe the community can try it out

**Review by sayakpaul — APPROVED** [review]
Dope! Thank you so much for seeing this through, Yiyi!

**Review by patrickvonplaten — APPROVED** [review]
